### PR TITLE
Replace `&mut self` with `&self` in AES peripheral

### PR DESCRIPTION
**Pull request by jorgeig-space**
The mutable references to `self` are not needed in the HAL methods, but make it hard to integrate the peripheral with upper layers (e.g. [the `cipher` traits](https://docs.rs/cipher/0.3.0/cipher/trait.BlockEncrypt.html)).

### CODE REVIEW

**Comment by jorgeig-space** [comment]
Not sure why GitHub shows so many commits when I keep our GitLab repo always synced to yours... but at least the resulting changes are as expected.

**Comment by jorgeig-space** [comment]
Usage example with those traits: https://github.com/jorgeig-space/lorawan-wl/blob/feature/hardware_aes/src/lorawan_crypto.rs

**Comment by newAM** [comment]
I think this would break the guarantees provided by RTIC?  In RTIC it is possible for multiple tasks to get access to a `&self` reference: https://rtic.rs/dev/book/en/by-example/resources.html?highlight=mutabl#only-shared---access

If a higher-priority interrupt occurs during `set_key` (for example), and the interrupting task uses the AES peripheral it would result in undefined behavior of the output plaintext/ciphertext for the interrupted task.

**Comment by jorgeig-space** [comment]
**TL;DR**: this would _enable_ the user to break the guarantees by having only shared access. They would have to specifically declare the driver as only being accessible via shared reference, bypassing the locks. On the other hand, keeping the _self references_ as `&mut` would prevent the use of the peripheral by any of the existing crypto code that uses the `cipher` traits.

----

AFAIU the manual refers to the way the reference is created in RTIC. E.g., you can declare this:
```{Rust}
struct Resources {
        #[init([0;256])]
        buffer_tx: [u8; 256],
        #[init([0;256])]
        buffer_rx: [u8; 256],
        aesd: hal::aes::Aes,
    }
```

What you get in the task's Context is, e.g., a `&mut hal::aes::Aes`, which you lock for your task. **If the user decides not to have locks** (e.g., because there won't be modifications to the resource), they can declare it in the `resources` section of the task as a shared reference; e.g.:

```{Rust}
#[task(priority = 2, resources = [&buffer_rx, rcc], spawn = [lorawan_response])]
```

In this case, RTIC gives you a shared ref to `buffer_rx` **everywhere** and a mutable reference to `rcc`, but this was a user choice and not caused by how the resources are built.

In the case of drivers, just because the driver doesn't need a mutable reference to itself, the user should not immediately remove the locks from RTIC. However, having the mutable references impedes the use of the peripheral for more complicated crypto code or as a replacement for software implementations, because most (all?) of those require implementation of the `cipher` traits.

**Comment by newAM** [comment]
The cipher traits require a separate structure because they carry the key in the structure, would it be possible to wrap the `Aes` struct in a `RefCell` to make it accessible behind a `&self` reference?  Alternatively `UnsafeCell` for a zero-cost wrapper if `RefCell` is not performant enough?

You are right that this is hard to misuse with RTIC, but I am not familiar enough with the other RTOS' in the ecosystem to know if this could be easily misused elsewhere.  I am happy to investigate (exploring other RTOS' is on my TODO list for [arewertosyet.com](https://arewertosyet.com/)) if `Cell`'s do not work out for this.

**Comment by jorgeig-space** [comment]
Right, that's a good solution. I totally forgot about RefCell. Thanks!

Let me try it and come back, but I think it should be fine and we can keep
the semantics of the mutable self-reference in the HAL (which makes total
sense, the peripheral *is* modifying itself!).

**Comment by jorgeig-space** [comment]
Works like a charm. No need for this PR.
